### PR TITLE
runtime: document fairness guarantees and current behavior

### PR DESCRIPTION
**Pull request by Darksonn**
Closes: #6049

### CODE REVIEW

File: tokio/src/runtime/mod.rs [hunk 1]
```diff
@@ -172,6 +172,128 @@
...
+//! timers. It does this by periodically checking whether there are any IO
+//! resources or timers that are ready, and notifying the relevant task so that
+//! it will be polled.
+//!
+//! These checks are performed periodically between polling tasks. Under the
+//! same assumptions as the previous fairness guarantee, Tokio guarantees that
+//! it will notify tasks with an IO or timer event within some maximum number of
+//! time units.
+//!
+//! ## Current behavior (current thread runtime)
```
Comment (hds) [inline tokio/src/runtime/mod.rs:227]: Reading this title felt to me like it implied that there would be a "future behavior" section as well.

Given that the first paragraph calls out that this behavior may change, this title could be simplied to "Current thread runtime behavior". The same for the multi thread runtime heading below.

**Review by hds — APPROVED** [review]
A couple of nits. Take them or leave them as you wish.

File: tokio/src/runtime/mod.rs (hunk 1 above)
Comment (Darksonn) [reply tokio/src/runtime/mod.rs:227]: By using the word "current", I meant to emphasize that this section only describes how it works today, and not necessarily in the future.

Comment (carllerche) [reply tokio/src/runtime/mod.rs:227]: How about "Behavior at the time of writing"?

File: tokio/src/runtime/mod.rs [hunk 2]
```diff
@@ -172,6 +172,128 @@
...
+//!
+//! This section describes how the [current thread runtime] behaves today. This
+//! behavior may change in future versions of Tokio.
+//!
+//! The current thread runtime maintains two FIFO queues of tasks that are ready
+//! to be polled: the global queue and the local queue. The runtime will prefer
+//! to choose the next task to poll from the local queue, and will only pick a
+//! task from the global queue if the local queue is empty, or if it has picked
+//! a task from the local queue 31 times in a row. The number 31 can be
+//! changed using the [`global_queue_interval`] setting.
```
Comment (carllerche) [inline tokio/src/runtime/mod.rs:237]: The only reason this is still the case is nobody has applied the MT behavior here 😅

File: tokio/src/runtime/mod.rs [hunk 3]
```diff
@@ -172,6 +172,128 @@
...
+//!
+//! A multi thread runtime has a fixed number of worker threads, which are all
+//! created on startup. The multi thread runtime maintains one global queue, and
+//! a local queue for each worker thread. The local queue of a worker thread can
+//! fit at most 256 tasks. If more than 256 tasks are added to the local queue,
+//! then half of them are moved to the global queue to make space.
+//!
+//! The runtime will prefer to choose the next task to poll from the local
+//! queue, and will only pick a task from the global queue if the local queue is
+//! empty, or if it has picked a task from the local queue 61 times in a row.
```
Comment (carllerche) [inline tokio/src/runtime/mod.rs:261]: In the multi-threaded case, the global queue interval is based on a heuristic using the task mean-poll time. IIRC it targets 10ms intervals and does 10ms / mean-task-poll-time to figure out the number of tasks to poll.

File: tokio/src/runtime/mod.rs [hunk 4]
```diff
@@ -172,6 +172,128 @@
...
 //! [`Builder::enable_time`]: crate::runtime::Builder::enable_time
 //! [`Builder::enable_all`]: crate::runtime::Builder::enable_all
+//!
+//! # Detailed runtime behavior
+//!
+//! This section gives more details into how the Tokio runtime will schedule
+//! tasks for execution.
+//!
+//! At its most basic level, a runtime has a collection of tasks that need to be
+//! polled. It will repeatedly remove a task from that collection and poll it.
```
[the remark this answers is not in the document]
Comment (Darksonn) [reply tokio/src/runtime/mod.rs]: I changed it to use "schedule" everywhere.

File: tokio/src/runtime/mod.rs [hunk 5]
```diff
@@ -172,6 +172,128 @@
...
+//! Normally, tasks are polled only if they have notified the runtime by waking
+//! their waker. However, this is not guaranteed, and Tokio may poll tasks that
+//! have not notified the runtime under some circumstances. This is called a
+//! spurious wakeup.
+//!
+//! ## IO and timers
+//!
+//! Beyond just polling tasks, the runtime must also manage IO resources and
+//! timers. It does this by periodically checking whether there are any IO
+//! resources or timers that are ready, and notifying the relevant task so that
```
[the remark this answers is not in the document]
Comment (Darksonn) [reply tokio/src/runtime/mod.rs]: I changed this to "wake" everywhere.

**Pull request merged**